### DISCUSSION TITLE
[Feature]: Support Wasm loader in ESBuild

### DISCUSSION THREAD

**Discussion by seanaye** [remix-run/remix#2752] [Proposals]
### What is the new or updated feature that you are suggesting?

There is currently no way to load Wasm files in a idiomatic way in Remix. You can use workarounds like copying the `.wasm` to the public folder and calling fetch on it within a useEffect. But this is not great DX, nor does it work serverside.

Current Workaround method:

```ts
  const [mod, setMod] = useState<InitOutput | null>(null);
  useEffect(() => {
    (async () => {
      const res = await fetch("wasm_bg.wasm");
      const wasm = await load(res);
      setMod(wasm);
    })();
}, [])
...
```

With the ESBuild wasm loader plugin added I believe the api could look like this
```ts
import wasm from "./wasm"
```

This obviously requires a top level await since Wasm loads asynchronously but thats what's so great about Remix! The component can Suspend while the wasm is loading. 

### Why should this feature be included?

Wasm is a well supported web standard thats being used more and more. Devs will need to use wasm in Remix applications. They are either going to add it in a hacky way like I did 🥸 or Remix could add support for it and provide amazing DX

**Comment by XYCaptain** [comment]
Need it!

**Reply by benwis** [reply]
This is a bit tricky, I have an esbuild plugin that will include imported wasm files in your build for both server and client, so that part is done. I'm just not sure how I want to handle the wrappers that pass data in and out. For Rust, wasm-bindgen will generate the wrapper functions necessary to make using it painless, but then I have to pass it through esbuild or mark it external

**Reply by benwis** [reply]
@seanaye How are you generating your WASM?

**Reply by seanaye** [reply]
Rust with wasm-bindgen

**Comment by brophdawg11** [comment]
This discussion is being closed as part of the Remix v3 repository cleanup.

Remix v1/v2 was [upstreamed into React Router v7](https://remix.run/blog/wake-up-remix), so if this proposal is still relevant, please reopen or continue the discussion in the [React Router](https://github.com/remix-run/react-router) repository.
